### PR TITLE
[RFC] Add uc_mmio_map for memory mapped devices

### PR DESCRIPTION
**Pull request by kitling**
This is an implementation of my second proposal for issue #565. I will admit, I am unfamiliar with the codebase, so there are likely many improvements that can be made.

I'll probably push a commit to my project, raspcorn, that uses this PR, soon.
### TODO
- [ ] Make an example, and perhaps some unit tests for this.

### CODE REVIEW

**Comment by kitling** [comment]
Latest commit of raspcorn uses this in a basic way. Any changes to the API added in this PR will be reflected in my application as I develop it.

@aquynh I'm sorry if I'm being impatient, or if you're reviewing this in your mailing list, but I never really got the hang of mailing lists... >_>

**Comment by aquynh** [comment]
Yes i do, but this is slow, so please be patient. Thanks for your
contribution!

**Comment by aquynh** [comment]
would you please add a sample code using this new API? something like `samples/sample_mmio.c` will be great.

File: uc.c [hunk 1]
```diff
@@ -1144,6 +1144,34 @@ uint32_t uc_mem_regions(uc_engine *uc, uc_mem_region **regions, uint32_t *count)
...
+
+    if (uc->mem_redirect) {
+        address = uc->mem_redirect(address);
+    }
+
+    res = mem_map_check(uc, address, size, UC_PROT_ALL);
+    if (res)
+        return res;
+
+    data = g_new(struct mmio_data, 1);
```
Comment (aquynh) [inline uc.c]: where are we going to free this memory? is this memory leaking?

Comment (kitling) [inline uc.c]: ... DOH. Can't believe that I forgot about this. It *is* leaking at the moment, isn't it? Need to: (A) remove the need for that struct. (B) be able to tell mmio apart from ram and free that memory when unmapping the mmio region. (C) create a seperate unmapping function.

You have any ideas on how to accomplish A? I don't mind changing the API to require 2 callbacks, for example, just depends on what you think is the best interface to present. I'll be happy that the API is there.

Comment (aquynh) [inline uc.c]: are you relying on uc_mem_unmap() to free the resource for you? if then, then uc_mem_unmap() must be aware of this allocated memory to release that.

**Comment by aquynh** [comment]
this adds a new API, however. another approach is to extend the existing API uc_mem_map(), so `perms` param accepts a new enum like UC_MMIO_MAP. so if you pass `UC_MMIO_MAP | UC_PROT_READ`, then your handler will be called upon reading that mmio area.

any objections?

**Comment by aquynh** [comment]
in this pull req, your API set the callback via a new API, which is not consistent with the existing API: we add & delete the callbacks via `uc_hook_add()` & `uc_hook_del()`, but not in the mapping functions.

i think the API change proposed above keep it simple and consistent.

**Comment by aquynh** [comment]
what do you have to say?

**Comment by kitling** [comment]
I apologize, I've gotten busy as of late, between attempting to test the project (NTRPi) that I've been working on, and school being active for me.

That said, I still want to get something in, but I'm not working on this at the moment. >_>

**Comment by lunixbochs** [comment]
I vote for MMIO extending the existing memory hook API. I might take a look at this.

**Comment by markyi370** [comment]
Has anyone fixed the memory leak? Waiting for this to be implemented.

**Comment by kitling** [comment]
Apologies once again. I haven't worked on this in... awhile. Just different project after different project. I ended up testing NTRPi on actual hardware, and found it to not be fast enough, so I have no real need for this anymore.

Since I don't find it likely that I'll come back to this, I'm closing the PR, as it has several issues.
 - It allocates memory without freeing it
 - It creates inconsistant new API
 - Even if it were a fine API, reading and writing should be seperate callbacks, anyway.
   - I think this could fix the memory leak, as iirc I only allocated memory so that I could squash the callbacks into the same function.

**Pull request closed without merging**